src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,487 @@
+	{
+		/* connection throttling disabled */
+		return;
+	}
+
+	if (SessionLocalReservedConnectionCounters == NULL)
+	{
+		/*
+		 * This is unexpected as SessionLocalReservedConnectionCounters hash table is
+		 * created at startup. Still, let's be defensive.
+		 */
+		ereport(ERROR, (errmsg("unexpected state for session level reserved "
+							   "connection counter hash")));
+	}
+
+	/*
+	 * We sort the workerList because adaptive connection management
+	 * (e.g., OPTIONAL_CONNECTION) requires any concurrent executions
+	 * to wait for the connections in the same order to prevent any
+	 * starvation. If we don't sort, we might end up with:
+	 *      Execution 1: Get connection for worker 1, wait for worker 2
+	 *      Execution 2: Get connection for worker 2, wait for worker 1
+	 *
+	 *  and, none could proceed. Instead, we enforce every execution establish
+	 *  the required connections to workers in the same order.
+	 */
+	nodeList = SortList(nodeList, CompareWorkerNodes);
+
+	char *databaseName = get_database_name(MyDatabaseId);
+
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (ConnectionAvaliableToNode(workerNode->workerName, workerNode->workerPort,
+									  CurrentUserName(), databaseName))
+		{
+			/*
+			 * The same user has already an active connection for the node. It
+			 * means that the execution can use the same connection, so reservation
+			 * is not necessary.
+			 */
+			continue;
+		}
+
+		if (HasAlreadyReservedConnection(workerNode->workerName, workerNode->workerPort,
+										 MyDatabaseId))
+		{
+			/*
+			 * We already locally reserved a connection for this node. So, it is not
+			 * necessary to reserve again.
+			 *
+			 * One caveat here is that we are ignoring for which user we reserved the
+			 * connection. It is not important as Citus doesn't support switching
+			 * between users (https://github.com/citusdata/citus/pull/3869) and
+			 * we're clearing the reserved connections at the end of the transaction.
+			 */
+			continue;
+		}
+
+		/*
+		 * We are trying to be defensive here by ensuring that the required hash
+		 * table entry can be allocated. The main goal is that we don't want to be
+		 * in a situation where shared connection counter is incremented but not
+		 * the local reserved counter due to out-of-memory.
+		 *
+		 * Note that shared connection stats operate on the shared memory, and we
+		 * pre-allocate all the necessary memory. In other words, it would never
+		 * throw out of memory error.
+		 */
+		AllocateReservedConectionEntry(workerNode->workerName, workerNode->workerPort,
+									   MyDatabaseId);
+
+		/*
+		 * Increment the shared counter, we may need to wait if there are
+		 * no space left.
+		 */
+		WaitLoopForSharedConnection(workerNode->workerName, workerNode->workerPort);
+
+		/* locally mark that we have one connection reserved */
+		IncrementReservedConnection(workerNode->workerName, workerNode->workerPort,
```

Comment:
Could AllocateReservedConectionEntry return the entry, such that we only need to do entry->reservedConnectionCount++ ?